### PR TITLE
Deserialize Species

### LINKED ISSUE #1016: Implement deserialization for Species
As part of #1013 implement the `deserialize()` method for Species. `SpeciesIntra` found in `speciesintra.h` and `species_intra.cpp` provide useful references on how to deal with bonds, angles, torsions and impropers.

This will require the deserialization of the following elements:
- [x] Atom
- [ ] Bond
- [ ] Angle
- [ ] Torsion
- [ ] Improper
- [ ] Isotopologue

### PR DESCRIPTION
**Pull request by IVlaD17**
Closes #1016 

Dependent on #1008

### CODE REVIEW

File: src/classes/speciesatom.cpp [hunk 1]
```diff
@@ -468,3 +468,24 @@ SerialisedValue SpeciesAtom::serialise() const
     atom["type"] = atomType_->name().data();
     return atom;
 }
+void SpeciesAtom::deserialise(SerialisedValue &node)
+{
+    if (!node["index"].is_uninitialized())
```
Comment (rprospero) [inline src/classes/speciesatom.cpp:473]: contains

File: src/classes/species.cpp [hunk 2]
```diff
@@ -221,3 +222,57 @@ SerialisedValue Species::serialise() const
...
+void Species::deserialise(SerialisedValue &node, CoreData &coreData)
+{
+    std::vector tomlAtoms = toml::find(node, "atoms").as_array();
+    for (auto tomlAtom : tomlAtoms)
+        atoms_.emplace_back().deserialise(tomlAtom);
+
+    Serialisable::toVector(node, "bonds",
+                           [this, &coreData](SerialisedValue &bond)
+                           {
+                               if (!bond["i"].is_uninitialized() && !bond["j"].is_uninitialized())
```
Comment (rprospero) [inline src/classes/species.cpp:235]: Remove checks and just throw exception

File: src/classes/masters.h [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+#include "classes/speciesbond.h"
+#include "classes/speciesimproper.h"
+#include "classes/speciestorsion.h"
+
+class Masters : public Serialisable
+{
+    public:
+    Masters() = default;
+    // Master Bond parameters for Species
+    std::vector<std::shared_ptr<MasterBond>> masterBonds;
```
Comment (rprospero) [inline src/classes/masters.h:17]: remove redundant `master` name

File: src/classes/speciesintra.h [hunk 4]
```diff
@@ -154,4 +155,66 @@ template <class Intra, class Functions> class SpeciesIntra
...
+        if (node.contains("parameters"))
+        {
+            std::vector<std::string> parameters = Functions::parameters(interactionForm());
+            std::vector<double> values;
+            std::transform(parameters.begin(), parameters.end(), std::back_inserter(values),
+                           [&node](const auto param) { return node["parameters"][param].as_floating(); });
+            setInteractionFormAndParameters(interactionForm(), values);
+        }
+    }
+    SerialisedValue serialise() const override
```
Comment (rprospero) [inline src/classes/speciesintra.h:199]: Add Comment

File: src/classes/speciesintra.h [hunk 5]
```diff
@@ -154,4 +155,66 @@ template <class Intra, class Functions> class SpeciesIntra
...
+        if (node.contains("parameters"))
+        {
+            std::vector<std::string> parameters = Functions::parameters(interactionForm());
+            std::vector<double> values;
+            for (auto parameter : parameters)
+                values.push_back(node["parameters"][parameter].as_floating());
+            setInteractionFormAndParameters(interactionForm(), values);
+        }
+    }
+    void deserialise(SerialisedValue &node) override
```
Comment (rprospero) [inline src/classes/speciesintra.h:183]: Add Comment

File: src/classes/speciessite.cpp [hunk 6]
```diff
@@ -434,3 +434,30 @@ SerialisedValue SpeciesSite::serialise() const
     site["originMassWeighted"] = originMassWeighted_;
     return site;
 }
+
+void SpeciesSite::deserialise(SerialisedValue &node)
```
Comment (rprospero) [inline src/classes/speciessite.cpp:438]: Use helper functions

File: src/main/io.cpp [hunk 7]
```diff
@@ -191,6 +193,22 @@ bool Dissolve::loadInput(std::string_view filename)
         setInputFilename(filename);
     }
 
+    if (toml_testing_flag)
```
Comment (rprospero) [inline src/main/io.cpp:196]: Remove testing code

**Review by rprospero — COMMENTED** [review]
Some thoughts before this gets reviewed

File: src/classes/speciesintra.h [hunk 8]
```diff
@@ -154,4 +155,66 @@ template <class Intra, class Functions> class SpeciesIntra
...
+            result["form"] = "@" + std::string(masterTerm_->name());
+        else
+            result["form"] = Functions::forms().keyword(interactionForm());
+
+        std::vector<double> values = interactionPotential().parameters();
+        if (!values.empty())
+        {
+            SerialisedValue parametersNode;
+            std::vector<std::string> parameters = Functions::parameters(interactionForm());
+            for (int parameterIndex = 0; parameterIndex < values.size(); parameterIndex++)
```
Comment (trisyoungs) [inline src/classes/speciesintra.h:213]:
```suggestion
            for (auto parameterIndex = 0; parameterIndex < values.size(); ++parameterIndex)
```

File: src/classes/speciestorsion.cpp [hunk 9]
```diff
@@ -596,7 +597,7 @@ double SpeciesTorsion::force(double angleInDegrees) const
 // Express as a tree node
 SerialisedValue SpeciesTorsion::serialise() const
 {
-    SerialisedValue torsion;
+    SerialisedValue torsion = SpeciesIntra<SpeciesTorsion, TorsionFunctions>::serialise();
```
Comment (trisyoungs) [inline src/classes/speciestorsion.cpp:600]:
```suggestion
    auto torsion = SpeciesIntra<SpeciesTorsion, TorsionFunctions>::serialise();
```

**Review by trisyoungs — CHANGES_REQUESTED** [review]
Looking good!  Just minor comments - bit of `auto` and stuff like that.

File: src/classes/configuration.h [hunk 10]
```diff
@@ -232,6 +232,8 @@ class Configuration : public Serialisable
     public:
     // Write through specified LineParser
     bool serialise(LineParser &parser) const;
+    // Read through specified LineParser
+    bool read(LineParser &parser, const std::vector<std::unique_ptr<Species>> &availableSpecies, double pairPotentialRange);
```
[the remark this answers is not in the document]
Comment (rprospero) [reply src/classes/configuration.h:236]: Good catch! That's a vestigal bit of code from an *old* commit.  The function doesn't even have an implementation.  I've removed it now.

File: src/classes/isotopologue.h [hunk 11]
```diff
@@ -22,6 +23,7 @@ class Isotopologue : public Serialisable
 {
     public:
     Isotopologue();
+    Isotopologue(std::string name);
```
[the remark this answers is not in the document]
Comment (rprospero) [reply src/classes/isotopologue.h:26]: That is the better way.

File: src/classes/masters.h [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+    public:
+    Masters() = default;
+    // Master Bond parameters for Species
+    std::vector<std::shared_ptr<MasterBond>> bonds;
+    // Master Angles parameters for Species
+    std::vector<std::shared_ptr<MasterAngle>> angles;
+    // Master Torsions parameters for Species
+    std::vector<std::shared_ptr<MasterTorsion>> torsions;
+    // Master Improper parameters for Species
+    std::vector<std::shared_ptr<MasterImproper>> impropers;
```
[the remark this answers is not in the document]
Comment (rprospero) [reply src/classes/masters.h:23]: I thought pretty hard about this.  What I've done is make the Masters class a private inner class of CoreData.  This means that we don't need the ceremony of getters and setters, but we've also indicated that this class should only ever be instantiated within the CoreData class, where it is a private member, giving us the encapsulation we would expect.

File: src/classes/masters.h [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+    SerialisedValue serialise() const override;
+    void deserialise(SerialisedValue &node) override;
```
[the remark this answers is not in the document]
Comment (rprospero) [reply src/classes/masters.h]: I've added the comment to the Masters new location.

File: src/classes/species.h [hunk 14]
```diff
@@ -30,6 +30,7 @@ class Species : public Serialisable
 {
     public:
     Species();
+    Species(std::string name);
```
[the remark this answers is not in the document]
Comment (rprospero) [reply src/classes/species.h:33]: Agreed

**Review by trisyoungs — APPROVED** [review]
:+1:

**Pull request merged**